HW1/P4/P4.txt

```diff
@@ -0,0 +1,18 @@
+# Your discussion here
+My graph representation was an rdd of edge tuples: (nodeA, nodeB)
+I first had an adjacency list but I found it hard and slow to manipulate list values in spark so I switched to the edge tuple representation and it’s much faster.
+
+Some optimizations to reduce shuffle costs:
+	- the join in my bfs for loop is done from the graph (graph.join(queue, N)) because the graph is a bigger rdd than the queue.
+	- I keep things co-partitioned whenever I can
+	- I reduce the size of my graph in every iteration by removing nodes I’ve already touched, which should make the joins much faster
+	- I mapValues instead of map whenever I can to avoid repartitioning
```

Comment:
Thanks for describing the optimizations you made. I considered doing some of them for my BFS (particularly the third one), but found that they actually slowed my code down. Perhaps the graph representation was key. Overall, very clean and concise BFS!